### PR TITLE
Import rework

### PR DESCRIPTION
**Pull request by stefnotch**

### CODE REVIEW

File: linker/packages/mini-parse/src/ParserCombinator.ts [hunk 1]
```diff
@@ -177,6 +242,33 @@ export function not(arg: CombinatorArg): Parser<true> {
   return notParser;
 }
 
+/** return true if the provided parser _doesn't_ match
+ * does not consume any tokens */
+export function span<P extends CombinatorArg>(
```
Comment (mighdoll) [inline linker/packages/mini-parse/src/ParserCombinator.ts:247]: .mapValue(), then .span() is pretty much just what map() does now. So not sure span() carries its weight... 

also, if we don want span(), I wonder if it would be better as a combinator, e.g. span(foo, bar). rather than a .span()..

File: linker/packages/linker/src/ImportGrammar.ts [hunk 2]
```diff
@@ -1,156 +1,120 @@
...
-    importTokens,
-    seq("import", fullPath, opt(";"), eolf)     .collect(importElem),
+    mainTokens,
+    delimited(
+      "import",
+      seq(
+        or(import_relative, import_package),
+        or(import_collection, import_path, item_import),
+      ).mapValue(v => {
+        if (v[1] instanceof ImportStatement) {
```
Comment (mighdoll) [inline linker/packages/linker/src/ImportGrammar.ts:85]: I was kinda liking the style (that you suggested AFAIR!) of moving these functions out of the grammar and putting them in a separate file. 

Could try that later.

File: linker/packages/mini-parse/src/ParserCombinator.ts [hunk 3]
```diff
@@ -107,6 +108,70 @@ export function seq<P extends CombinatorArg[]>(...args: P): SeqParser<P> {
...
 
+/** Parse two values, and discard the first value
+ * @return the second value, or null if any parser fails */
+export function preceded<P extends CombinatorArg>(
+  ignoredArg: CombinatorArg,
+  arg: P,
+): ParserFromArg<P> {
+  const ignored = parserArg(ignoredArg);
+  const p = parserArg(arg);
+  const seqParser: ParserFromArg<P> = parser("seq", (ctx: ParserContext) => {
```
Comment (mighdoll) [inline linker/packages/mini-parse/src/ParserCombinator.ts:119]: why not actually use seq()?

File: linker/packages/mini-parse/src/ParserCombinator.ts [hunk 4]
```diff
@@ -107,6 +108,70 @@ export function seq<P extends CombinatorArg[]>(...args: P): SeqParser<P> {
...
+    if (ignoredResult === null) return null;
+    return result;
+  }).collect({ before: pushOpenArray, after: closeArray });
+
+  trackChildren(seqParser, ignored, p);
+
+  return seqParser;
+}
+
+/** Parse two values, and discard the first value
```
Comment (mighdoll) [inline linker/packages/mini-parse/src/ParserCombinator.ts:151]: fix comment

File: linker/packages/mini-parse/src/ParserCombinator.ts [hunk 5]
```diff
@@ -107,6 +108,70 @@ export function seq<P extends CombinatorArg[]>(...args: P): SeqParser<P> {
...
+ * @return the second value, or null if any parser fails */
+export function delimited<P extends CombinatorArg>(
+  ignoredArg1: CombinatorArg,
+  arg: P,
+  ignoredArg2: CombinatorArg,
+): ParserFromArg<P> {
+  const ignored1 = parserArg(ignoredArg1);
+  const p = parserArg(arg);
+  const ignored2 = parserArg(ignoredArg2);
+  const seqParser: ParserFromArg<P> = parser("seq", (ctx: ParserContext) => {
```
Comment (mighdoll) [inline linker/packages/mini-parse/src/ParserCombinator.ts:161]: why not call it "delimited"?

File: linker/packages/mini-parse/src/ParserCombinator.ts (hunk 3 above)
Comment (stefnotch) [reply linker/packages/mini-parse/src/ParserCombinator.ts:119]: Because I wanted the semantics of also throwing away the tags from the ignored parser. Using `seq` was what I did for my first implementation, until I ran into that.

File: linker/packages/mini-parse/src/ParserCombinator.ts (hunk 5 above)
Comment (stefnotch) [reply linker/packages/mini-parse/src/ParserCombinator.ts:161]: Copy pasta mistake, sorry

File: linker/packages/linker/src/ImportGrammar.ts [hunk 6]
```diff
@@ -1,156 +1,120 @@
...
+        (v): ImportElem => ({
+          kind: "import",
+          contents: [],
+          imports: v.value,
+          start: v.span[0],
+          end: v.span[1],
+        }),
+      )
+      .ptag("owo")
+      .collect(importElem),
```
Comment (mighdoll) [inline linker/packages/linker/src/ImportGrammar.ts:107]: It's very interesting how you've combined parse-synchronous map() and post parse collect().
totally legit. 
Over the next weeks, lets figure out how/when we ought use map() vs. collect().

File: linker/packages/mini-parse/src/ParserCombinator.ts (hunk 1 above)
Comment (stefnotch) [reply linker/packages/mini-parse/src/ParserCombinator.ts]: I removed the `span` after realizing that it's bugged

File: linker/packages/linker/src/ImportGrammar.ts (hunk 2 above)
Comment (stefnotch) [reply linker/packages/linker/src/ImportGrammar.ts:91]: Let's try that later, I'm not entirely sure what the cleanest approach here would be.

File: linker/packages/linker/src/ImportTree.ts [hunk 7]
```diff
@@ -1,19 +1,42 @@
-export class ImportTree {
-  /** segments in path order */
-  constructor(public segments: PathSegment[]) {}
+/**
```
Comment (mighdoll) [inline linker/packages/linker/src/ImportTree.ts:1]: rename file

File: linker/packages/linker/src/ImportTree.ts [hunk 8]
```diff
@@ -1,19 +1,42 @@
...
+ * An import statement, which is tree shaped.
+ * `import foo::bar::{baz, cat as neko};
```
Comment (mighdoll) [inline linker/packages/linker/src/ImportTree.ts:3]: like the running example

File: linker/packages/linker/src/ImportTree.ts [hunk 9]
```diff
@@ -1,19 +1,42 @@
...
+ */
+export class ImportStatement {
+  constructor(
+    public segments: ImportSegment[],
+    public finalSegment: ImportCollection | ImportItem,
```
Comment (mighdoll) [inline linker/packages/linker/src/ImportTree.ts:8]: nice idea to collect the finalSegment during parsing

once upon a time, I allowed {} collections in the middle of an import path. 
probably that won't be allowed in the future, right?

File: linker/packages/linker/src/ImportGrammar.ts [hunk 10]
```diff
@@ -1,156 +1,125 @@
...
+  ImportCollection,
+  ImportItem,
+  ImportSegment,
+  ImportStatement,
+} from "./ImportTree.js";
+import { ImportElem } from "./AbstractElems.js";
+import { importElem } from "./WESLCollect.js";
+
+const wordToken = kind(mainTokens.ident)
+  // TODO: Remove this temporary hack to make the tests pass once we have a better tokenizer
```
Comment (mighdoll) [inline linker/packages/linker/src/ImportGrammar.ts:33]: why not give super (and package?) their own token types? Seems like it could be handled there

File: linker/packages/linker/src/StandardTypes.ts [hunk 11]
```diff
@@ -37,6 +37,183 @@ export const textureStorageTypes = `
   texture_storage_3d
 `;
 
+/** https://www.w3.org/TR/WGSL/#keyword-summary followed by https://www.w3.org/TR/WGSL/#reserved-words */
+export const keywords_and_reserved_words = [
```
Comment (mighdoll) [inline linker/packages/linker/src/StandardTypes.ts:41]: I can imagine this will be useful, but seems unused so far. What use do you have in mind? (Or did I miss it?)

File: linker/packages/linker/src/StandardTypes.ts [hunk 12]
```diff
@@ -37,6 +37,183 @@ export const textureStorageTypes = `
...
+  "break",
+  "case",
+  "const",
+  "const_assert",
+  "continue",
+  "continuing",
+  "default",
+  "diagnostic",
+  "discard",
+  "else",
```
Comment (mighdoll) [inline linker/packages/linker/src/StandardTypes.ts:52]: can format more snuggly, for density, like the other lists?

File: linker/packages/linker/src/test/ParseWESL.test.ts [hunk 13]
```diff
@@ -960,64 +960,59 @@ test("var foo: vec2<f32 >= vec2( 0.5, -0.5);", ctx => {
   `);
 });
 
-test("import ./a/b/c", ctx => {
+test("import a::b::c;", ctx => {
```
Comment (mighdoll) [inline linker/packages/linker/src/test/ParseWESL.test.ts:963]: yea! moving forward to the new syntax!

File: linker/packages/mini-parse/src/TokenMatcher.ts [hunk 14]
```diff
@@ -39,7 +41,7 @@ export function tokenMatcher<T extends Record<string, string | RegExp>>(
   const groups: string[] = Object.keys(matchers);
   let src: string;
   // cache of tokens by position, so we don't have to reparse after backtracking
-  const cache = new Cache<number, Token>(5);
+  const cache = new Cache<number, { token: Token; span: Span }>(5);
```
Comment (mighdoll) [inline linker/packages/mini-parse/src/TokenMatcher.ts:44]: 5, carefully chosen by random dice roll

File: linker/packages/mini-parse/src/Span.ts [hunk 15]
```diff
@@ -0,0 +1 @@
+export type Span = readonly [number, number];
```
Comment (mighdoll) [inline linker/packages/mini-parse/src/Span.ts:1]: document start and end (I had to guess whether start and length or start and end)

File: linker/packages/mini-parse/src/ParserCombinator.ts [hunk 16]
```diff
@@ -107,6 +134,110 @@ export function seq<P extends CombinatorArg[]>(...args: P): SeqParser<P> {
   return seqParser as SeqParser<P>;
 }
 
+/** Parse a sequence of parsers
+ * @return an array of all parsed results, or null if any parser fails */
+export function seqObj<P extends { [key: string]: CombinatorArg }>(
```
Comment (mighdoll) [inline linker/packages/mini-parse/src/ParserCombinator.ts:139]: more explanation, tags presumably

**Review by mighdoll — COMMENTED** [review]
wallaby and vitest are failing when I try it (after pnpm i). Does it work for you?
I get:
Tests  3 failed | 330 passed | 4 skipped (337)

**Comment by mighdoll** [comment]
awesome to see the import cases all passing

**Comment by mighdoll** [comment]
I'm surprised there aren't new import cases despite some spec changes, etc. Don't hold those corner cases in! add some cases while imports are relatively fresh in your mind? we'll use 'em a lot and for other tools too. (obviously can be separate commit)

**Review by mighdoll — COMMENTED** [review]
tests need to pass of course. Otherwise fix or defer minor comments and lgtm.

**Comment by stefnotch** [comment]
> wallaby and vitest are failing when I try it (after pnpm i). Does it work for you? I get: Tests 3 failed | 330 passed | 4 skipped (337)

Apparently running vitest from the root is still broken for me
```
D:\Coding\GitHub\Tools\wesl-js\linker> pnpm run test --ui                                                                                                                                                                                                                                   02/03/2025 02:22:50 PM

> root@ test D:\Coding\GitHub\Tools\wesl-js\linker
> vitest "--ui"

failed to load config from D:\Coding\GitHub\Tools\wesl-js\linker\packages\plugin-test\vite.config.ts

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Startup Error ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
Error [ERR_MODULE_NOT_FOUND]: Cannot find module 'D:\Coding\GitHub\Tools\wesl-js\linker\packages\plugin\node_modules\wesl\dist\index.js' imported from D:\Coding\GitHub\Tools\wesl-js\linker\packages\plugin\dist\chunk-2PLORNUK.js
    at finalizeResolution (node:internal/modules/esm/resolve:264:11)
    at moduleResolve (node:internal/modules/esm/resolve:917:10)
    at defaultResolve (node:internal/modules/esm/resolve:1130:11)
    at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:396:12)
    at ModuleLoader.resolve (node:internal/modules/esm/loader:365:25)
    at ModuleLoader.getModuleJob (node:internal/modules/esm/loader:240:38)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:85:39)
    at link (node:internal/modules/esm/module_job:84:36) {
  code: 'ERR_MODULE_NOT_FOUND',
  url: 'file:///D:/Coding/GitHub/Tools/wesl-js/linker/packages/plugin/node_modules/wesl/dist/index.js'
}
```

File: linker/packages/linker/src/ImportTree.ts (hunk 9 above)
Comment (stefnotch) [reply linker/packages/linker/src/ImportTree.ts:8]: Yeah, I think there's no value in allowing it in the middle.

File: linker/packages/linker/src/ImportTree.ts (hunk 8 above)
Comment (stefnotch) [reply linker/packages/linker/src/ImportTree.ts:3]: Thank you :)

File: linker/packages/linker/src/ImportTree.ts (hunk 7 above)
Comment (stefnotch) [reply linker/packages/linker/src/ImportTree.ts:1]: Good catch, will fix

File: linker/packages/linker/src/ImportGrammar.ts (hunk 10 above)
Comment (stefnotch) [reply linker/packages/linker/src/ImportGrammar.ts:33]: We can't do that just yet, because `superfoo` should be a single token. I could do the hacks that we discussed on Discord, but no value in doing so if I could just wait for the better tokenizer to be merged in.

File: linker/packages/linker/src/StandardTypes.ts (hunk 11 above)
Comment (stefnotch) [reply linker/packages/linker/src/StandardTypes.ts:41]: The better tokenizer will use that. Although it moves this to a different file. Lemme remove this then.

File: linker/packages/mini-parse/src/TokenMatcher.ts (hunk 14 above)
Comment (stefnotch) [reply linker/packages/mini-parse/src/TokenMatcher.ts:44]: Haha indeed!
I copied it from the other place that had the cache.

File: linker/packages/mini-parse/src/Span.ts (hunk 15 above)
Comment (stefnotch) [reply linker/packages/mini-parse/src/Span.ts:1]: Whoops, sorry! I'll document that!

File: linker/packages/mini-parse/src/ParserCombinator.ts (hunk 16 above)
Comment (stefnotch) [reply linker/packages/mini-parse/src/ParserCombinator.ts:139]: I added a bit more documentation, but I think longer term, I should add an example to it.

I just wish that Typescript had [Python's doctest](https://docs.python.org/3/library/doctest.html) or [Rust's doctest](https://doc.rust-lang.org/rustdoc/write-documentation/documentation-tests.html).

Oha!
Actually, [Deno can do that](https://docs.deno.com/runtime/reference/documentation/)! ✨

**Comment by stefnotch** [comment]
This is ready to merge. While the spec has changed, it hasn't changed in a significant enough way to actually impact this.

(I'll update the import parser to look 100% like the spec in a future PR)

**Comment by stefnotch** [comment]
I'm merging this
